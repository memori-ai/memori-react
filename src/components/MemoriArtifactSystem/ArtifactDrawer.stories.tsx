// ArtifactSystem.stories.tsx
import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import { ArtifactProvider } from './context/ArtifactContext';
import ArtifactHandler from './components/ArtifactHandler/ArtifactHandler';
import ArtifactDrawer from './components/ArtifactDrawer/ArtifactDrawer';

// Mock chat component for testing
const MockChat = ({ messages, isChatlogPanel }: { messages: any[], isChatlogPanel?: boolean }) => (
  <div style={{ padding: '20px', background: '#f5f5f5', minHeight: '400px' }}>
    {messages.map((message, index) => (
      <div key={index} style={{ marginBottom: '20px' }}>
        <div style={{ 
          background: message.fromUser ? '#007bff' : '#fff', 
          color: message.fromUser ? '#fff' : '#000',
          padding: '12px',
          borderRadius: '12px',
          marginBottom: '8px'
        }}>
          {message.text}
        </div>
        <ArtifactHandler message={message} isChatlogPanel={isChatlogPanel} />
      </div>
    ))}
  </div>
);

// Story decorator to provide context
const withArtifactProvider = (Story: any) => (
  <ArtifactProvider>
    <div style={{ fontFamily: 'Arial, sans-serif' }}>
      <Story />
      <ArtifactDrawer />
    </div>
  </ArtifactProvider>
);

const meta: Meta<typeof ArtifactHandler> = {
  title: 'Artifact System',
  component: ArtifactHandler,
  decorators: [withArtifactProvider],
  parameters: {
    layout: 'fullscreen',
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

// Sample artifacts for different mime types
const htmlArtifact = `<output class="memori-artifact" data-mimetype="html">
<!DOCTYPE html>
<html>
<head>
    <title>Sample Page</title>
    <style>
        body { font-family: Arial, sans-serif; padding: 20px; }
        .header { color: #333; border-bottom: 2px solid #007bff; }
        .content { margin-top: 20px; }
        .button { background: #007bff; color: white; padding: 10px 20px; border: none; border-radius: 5px; }
    </style>
</head>
<body>
    <h1 class="header">Welcome to My Page</h1>
    <div class="content">
        <p>This is a sample HTML page created as an artifact.</p>
        <button class="button" onclick="alert('Hello!')">Click me</button>
    </div>
</body>
</html>
</output>`;

const markdownArtifact = `<output class="memori-artifact" data-mimetype="markdown">
# Project Documentation

## Overview
This is a comprehensive guide for the project.

## Features
- **Easy to use**: Simple interface
- **Fast**: Optimized performance
- **Reliable**: Well tested

## Getting Started
1. Install dependencies
2. Run the application
3. Open your browser

## Code Example
\`\`\`javascript
function hello() {
    console.log("Hello World!");
}
\`\`\`

## Conclusion
This project helps you build amazing applications.
</output>`;

const cssArtifact = `<output class="memori-artifact" data-mimetype="css">
/* Modern CSS Grid Layout */
.container {
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
    gap: 20px;
    padding: 20px;
    max-width: 1200px;
    margin: 0 auto;
}

.card {
    background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
    border-radius: 12px;
    padding: 20px;
    color: white;
    box-shadow: 0 4px 20px rgba(0,0,0,0.1);
    transition: transform 0.3s ease;
}

.card:hover {
    transform: translateY(-5px);
    box-shadow: 0 8px 30px rgba(0,0,0,0.2);
}

.card h3 {
    margin-top: 0;
    font-size: 1.5em;
}

.card p {
    line-height: 1.6;
    opacity: 0.9;
}

@media (max-width: 768px) {
    .container {
        grid-template-columns: 1fr;
        padding: 10px;
    }
}
</output>`;

const javascriptArtifact = `<output class="memori-artifact" data-mimetype="javascript">
// Interactive Todo List Application
class TodoApp {
    constructor() {
        this.todos = [];
        this.init();
    }

    init() {
        this.createHTML();
        this.bindEvents();
    }

    createHTML() {
        const app = document.getElementById('app');
        app.innerHTML = \`
            <div class="todo-app">
                <h1>My Todo List</h1>
                <div class="input-section">
                    <input type="text" id="todoInput" placeholder="Add a new todo...">
                    <button id="addBtn">Add</button>
                </div>
                <ul id="todoList"></ul>
            </div>
        \`;
    }

    bindEvents() {
        document.getElementById('addBtn').addEventListener('click', () => this.addTodo());
        document.getElementById('todoInput').addEventListener('keypress', (e) => {
            if (e.key === 'Enter') this.addTodo();
        });
    }

    addTodo() {
        const input = document.getElementById('todoInput');
        const text = input.value.trim();
        
        if (text) {
            const todo = {
                id: Date.now(),
                text: text,
                completed: false
            };
            
            this.todos.push(todo);
            this.renderTodos();
            input.value = '';
        }
    }

    toggleTodo(id) {
        const todo = this.todos.find(t => t.id === id);
        if (todo) {
            todo.completed = !todo.completed;
            this.renderTodos();
        }
    }

    deleteTodo(id) {
        this.todos = this.todos.filter(t => t.id !== id);
        this.renderTodos();
    }

    renderTodos() {
        const list = document.getElementById('todoList');
        list.innerHTML = this.todos.map(todo => \`
            <li class="todo-item \${todo.completed ? 'completed' : ''}">
                <span onclick="app.toggleTodo(\${todo.id})">\${todo.text}</span>
                <button onclick="app.deleteTodo(\${todo.id})">Delete</button>
            </li>
        \`).join('');
    }
}

// Initialize the app
const app = new TodoApp();
</output>`;

const jsonArtifact = `<output class="memori-artifact" data-mimetype="json">
{
  "name": "user-profile-api",
  "version": "1.0.0",
  "description": "REST API for user profile management",
  "main": "server.js",
  "scripts": {
    "start": "node server.js",
    "dev": "nodemon server.js",
    "test": "jest",
    "build": "webpack --mode production"
  },
  "dependencies": {
    "express": "^4.18.0",
    "mongoose": "^6.0.0",
    "bcryptjs": "^2.4.3",
    "jsonwebtoken": "^8.5.1",
    "cors": "^2.8.5",
    "helmet": "^5.0.0",
    "morgan": "^1.10.0",
    "express-rate-limit": "^6.0.0"
  },
  "devDependencies": {
    "nodemon": "^2.0.15",
    "jest": "^27.5.1",
    "supertest": "^6.2.2"
  },
  "engines": {
    "node": ">=14.0.0"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/example/user-profile-api.git"
  },
  "keywords": ["api", "express", "mongodb", "authentication", "rest"],
  "author": "Your Name",
  "license": "MIT"
}
</output>`;

// Stories
export const HTMLArtifact: Story = {
  args: {},
  render: () => (
    <MockChat 
      messages={[
        { 
          text: `Here's a complete HTML page for your landing page:

${htmlArtifact}

This page includes modern styling and an interactive button. You can customize the colors and layout as needed.`,
          fromUser: false,
          timestamp: new Date().toISOString()
        }
      ]} 
    />
  ),
};

export const MarkdownDocumentation: Story = {
  args: {},
  render: () => (
    <MockChat 
      messages={[
        { 
          text: `I've created comprehensive project documentation for you:

${markdownArtifact}

This includes all the essential sections for a project README with proper formatting and structure.`,
          fromUser: false,
          timestamp: new Date().toISOString()
        }
      ]} 
    />
  ),
};

export const CSSStyles: Story = {
  args: {},
  render: () => (
    <MockChat 
      messages={[
        { 
          text: `Here's a modern CSS grid layout with hover effects:

${cssArtifact}

This creates responsive cards with beautiful gradients and smooth animations.`,
          fromUser: false,
          timestamp: new Date().toISOString()
        }
      ]} 
    />
  ),
};

export const JavaScriptApp: Story = {
  args: {},
  render: () => (
    <MockChat 
      messages={[
        { 
          text: `I've built a complete Todo List application in JavaScript:

${javascriptArtifact}

This includes full CRUD functionality with a clean class-based architecture.`,
          fromUser: false,
          timestamp: new Date().toISOString()
        }
      ]} 
    />
  ),
};

export const JSONConfiguration: Story = {
  args: {},
  render: () => (
    <MockChat 
      messages={[
        { 
          text: `Here's a complete package.json configuration for your API project:

${jsonArtifact}

This includes all necessary dependencies, scripts, and metadata for a production-ready Node.js API.`,
          fromUser: false,
          timestamp: new Date().toISOString()
        }
      ]} 
    />
  ),
};
export const MultipleArtifacts: Story = {
  args: {},
  render: () => (
    <MockChat 
      messages={[
        { 
          text: `I'll create both the HTML structure and CSS styling for your dashboard:

${htmlArtifact}

${cssArtifact}

This combination gives you a complete, styled web page ready for deployment.`,
          fromUser: false,
          timestamp: new Date().toISOString()
        }
      ]} 
    />
  ),
};

export const ConversationFlow: Story = {
  args: {},
  render: () => (
    <MockChat 
      messages={[
        { 
          text: "Can you create a simple HTML page for me?",
          fromUser: true,
          timestamp: new Date(Date.now() - 300000).toISOString()
        },
        { 
          text: `Of course! Here's a complete HTML page with modern styling:

${htmlArtifact}

Would you like me to modify anything about the design or add more functionality?`,
          fromUser: false,
          timestamp: new Date(Date.now() - 240000).toISOString()
        },
        { 
          text: "Now can you create the CSS to make it look more modern?",
          fromUser: true,
          timestamp: new Date(Date.now() - 180000).toISOString()
        },
        { 
          text: `Here's enhanced CSS with modern design patterns:

${cssArtifact}

This adds responsive grid layouts, smooth animations, and contemporary styling.`,
          fromUser: false,
          timestamp: new Date(Date.now() - 120000).toISOString()
        }
      ]} 
    />
  ),
};

export const NoArtifacts: Story = {
  args: {},
  render: () => (
    <MockChat 
      messages={[
        { 
          text: "Hello! How can I help you today?",
          fromUser: false,
          timestamp: new Date().toISOString()
        },
        { 
          text: "I need help with my project planning",
          fromUser: true,
          timestamp: new Date().toISOString()
        },
        { 
          text: "I'd be happy to help with project planning! What type of project are you working on? Are you looking for help with timeline management, resource allocation, task breakdown, or something else?",
          fromUser: false,
          timestamp: new Date().toISOString()
        }
      ]} 
    />
  ),
};

export const ChatLogPanelArtifact: Story = {
  args: {},
  render: () => (
    <MockChat
        isChatlogPanel={true}
      messages={[
        { 
          text: `Here's some code that has issues:
${htmlArtifact}`,
          fromUser: false,
          timestamp: new Date().toISOString()
        }
      ]} 
    />
  ),
};

export const InvalidArtifact: Story = {
  args: {},
  render: () => (
    <MockChat 
      messages={[
        { 
          text: `Here's some code that has issues:

<output class="memori-artifact" data-mimetype="html">
<h1>Broken HTML
<p>Missing closing tags
</output>

This artifact is too short and has invalid HTML, so it won't be displayed as an artifact.`,
          fromUser: false,
          timestamp: new Date().toISOString()
        }
      ]} 
    />
  ),
};

// CSS for the storybook stories
const styles = `
<style>
.artifact-handlers {
  margin: 10px 0;
}

.artifact-handler {
  display: flex;
  align-items: center;
  padding: 12px;
  border: 1px solid #ddd;
  border-radius: 8px;
  background: white;
  cursor: pointer;
  transition: all 0.2s ease;
  margin: 8px 0;
}

.artifact-handler:hover {
  background: #f8f9fa;
  border-color: #007bff;
  transform: translateY(-1px);
  box-shadow: 0 2px 8px rgba(0,123,255,0.15);
}

.artifact-handler-icon {
  font-size: 24px;
  margin-right: 12px;
}

.artifact-handler-info {
  flex: 1;
}

.artifact-handler-title {
  font-weight: 600;
  color: #333;
  margin-bottom: 4px;
}

.artifact-handler-meta {
  font-size: 12px;
  color: #666;
}

.artifact-handler-action {
  font-size: 18px;
  color: #007bff;
}

.artifact-drawer {
  position: fixed;
  top: 0;
  right: 0;
  width: 50%;
  height: 100vh;
  background: white;
  border-left: 1px solid #ddd;
  display: flex;
  flex-direction: column;
  z-index: 1000;
}

.artifact-drawer.fullscreen {
  width: 100%;
  left: 0;
}

.artifact-drawer-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 16px;
  border-bottom: 1px solid #ddd;
  background: #f8f9fa;
}

.artifact-drawer-header h3 {
  margin: 0;
  color: #333;
}

.artifact-drawer-actions button {
  background: none;
  border: 1px solid #ddd;
  padding: 8px;
  margin-left: 8px;
  border-radius: 4px;
  cursor: pointer;
}

.artifact-drawer-actions button:hover {
  background: #e9ecef;
}

.artifact-drawer-tabs {
  display: flex;
  background: #f8f9fa;
  border-bottom: 1px solid #ddd;
}

.artifact-drawer-tabs button {
  padding: 12px 24px;
  background: none;
  border: none;
  cursor: pointer;
  border-bottom: 2px solid transparent;
}

.artifact-drawer-tabs button.active {
  border-bottom-color: #007bff;
  color: #007bff;
}

.artifact-drawer-content {
  flex: 1;
  overflow: auto;
  padding: 16px;
}

.artifact-code {
  background: #f8f9fa;
  padding: 16px;
  border-radius: 4px;
  font-family: 'Monaco', 'Menlo', monospace;
  font-size: 14px;
  line-height: 1.5;
  margin: 0;
  overflow-x: auto;
}

.artifact-preview {
  height: 100%;
}

.artifact-preview iframe {
  width: 100%;
  height: 500px;
  border: 1px solid #ddd;
  border-radius: 4px;
}

@media (max-width: 768px) {
  .artifact-drawer {
    width: 100%;
    left: 0;
  }
}
</style>`;

// Add styles to document head
if (typeof document !== 'undefined') {
  const styleElement = document.createElement('style');
  styleElement.innerHTML = styles.replace(/<\/?style>/g, '');
  document.head.appendChild(styleElement);
}