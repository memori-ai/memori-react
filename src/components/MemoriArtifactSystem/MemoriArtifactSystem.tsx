/**
 * MemoriArtifactSystem Component
 * Main component that integrates the entire artifact system
 * Following the project's component patterns and design system
 */

import React, { useEffect, useCallback, useRef } from 'react';
import { useTranslation } from 'react-i18next';
import ArtifactDrawer from './components/ArtifactDrawer/ArtifactDrawer';
import ArtifactHandler from './components/ArtifactHandler/ArtifactHandler';
import {
  useArtifactSystem,
  useArtifactProcessor,
} from './hooks/useArtifactSystem';
import {
  ArtifactSystemActions,
  ArtifactSystemConfig,
  ArtifactSystemState,
  MemoriNewDialogStateEvent,
} from './types/artifact.types';

// Import CSS files
import { Message } from '@memori.ai/memori-api-client/dist/types';

export interface MemoriArtifactSystemProps {
  config: ArtifactSystemConfig;
  onArtifactCreated?: (artifact: any) => void;
  className?: string;
  message: Message;
}

const MemoriArtifactSystem: React.FC<MemoriArtifactSystemProps> = ({
  config,
  onArtifactCreated,
  message,
  className,
}) => {
  const { state, actions } = useArtifactSystem(config);
  const { processArtifactContent } = useArtifactProcessor();

  /**
   * Process new dialog state for artifacts
   */
  useEffect(() => {
    const handleMemoriNewDialogState = (message: string) => {
      if (!message || typeof message !== 'string') return;
      console.log('ðŸŽ¨ Processing message for artifacts...');

      // controllo
      // Process the message content to find and handle any artifacts
      // processArtifactContent returns true if artifacts were found and processed
      const processed = processArtifactContent(
        message, // The message text to scan for artifacts
        (content, mimeType, title) => {
          // Callback when artifact is found

          // Create a new artifact object with:
          // - Unique ID combining type, timestamp and random string
          // - Content and metadata from the artifact
          // - Type info from config or default fallback
          // - Generated or custom title
          // - Creation timestamp and size
          const artifact = {
            // Generate unique ID
            id: `artifact-${mimeType}-${Date.now()}-${Math.random()
              .toString(36)
              .substr(2, 9)}`,
            content, // The artifact content
            mimeType: mimeType as any,
            // Get type info from config or use default
            typeInfo: config.supportedMimeTypes?.[
              mimeType as keyof typeof config.supportedMimeTypes
            ] || {
              name: mimeType.toUpperCase(),
              icon: 'ðŸ“„',
              hasPreview: false,
              language: 'text',
              mimeType: 'text/plain',
            },
            // Use provided title or generate one from type info
            title:
              title ||
              `${
                config.supportedMimeTypes?.[
                  mimeType as keyof typeof config.supportedMimeTypes
                ]?.icon || 'ðŸ“„'
              } ${
                config.supportedMimeTypes?.[
                  mimeType as keyof typeof config.supportedMimeTypes
                ]?.name || mimeType.toUpperCase()
              } Artifact`,
            customTitle: title,
            timestamp: new Date(),
            size: content.length,
          };

          // Add the new artifact to the artifact system state
          actions.addArtifact(artifact);

          // If auto-open is enabled in config,
          // select/open the artifact after a short delay
          if (config.autoOpenArtifacts) {
            setTimeout(() => {
              actions.selectArtifact(artifact);
            }, 500);
          }

          // Call the onArtifactCreated callback if provided
          if (onArtifactCreated) {
            onArtifactCreated(artifact);
          }
        }
      );

      if (processed) {
        console.log('âœ… Artifacts processed successfully');
      }
    };
    if (!message.fromUser) {
      handleMemoriNewDialogState(message.translatedText || message.text);
    }
  }, [message]);

  return (
    <div className={className}>
      <ArtifactDrawer
        open={state.isDrawerOpen}
        onClose={actions.closeDrawer}
        artifact={state.currentArtifact}
        history={state.history}
        onSelectArtifact={actions.selectArtifact}
        onToggleFullscreen={actions.toggleFullscreen}
        isFullscreen={state.isFullscreen}
      />
      {state.history.map(artifact => (
        <ArtifactHandler
          onArtifactCreated={artifact => {
            actions.selectArtifact(artifact);
          }}
          key={artifact.id}
          artifact={artifact}
          content={artifact.content}
          mimeType={artifact.mimeType}
        />
      ))}
    </div>
  );
};

export default MemoriArtifactSystem;
