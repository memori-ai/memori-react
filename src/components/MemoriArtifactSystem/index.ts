/**
 * MemoriArtifactSystem - Main Export
 * Complete artifact system for Memori/Aisuru
 */

// Individual components
export { default as ArtifactDrawer } from './components/ArtifactDrawer/ArtifactDrawer';
export { default as ArtifactActions } from './components/ArtifactActions/ArtifactActions';
export { default as ArtifactPreview } from './components/ArtifactPreview/ArtifactPreview';
export { default as ArtifactHistory } from './components/ArtifactHistory/ArtifactHistory';
export { default as ArtifactHandler } from './components/ArtifactHandler/ArtifactHandler';

// Hooks
export { useArtifactSystem, useArtifactCreator, useArtifactProcessor } from './hooks/useArtifactSystem';

// Context
export { 
  ArtifactSystemProvider, 
  useArtifactSystemContext,
  ArtifactSystemContext 
} from './context/ArtifactSystemContext';

// Types
export type {
  ArtifactMimeType,
  ArtifactTypeInfo,
  ArtifactData,
  ArtifactHistoryEntry,
  ArtifactActionsProps,
  ArtifactPreviewProps,
  ArtifactHistoryProps,
  ArtifactHandlerProps,
  ArtifactSystemConfig,
  ArtifactSystemState,
  ArtifactSystemActions,
  ArtifactSystemHook,
  ArtifactEvent,
  MemoriNewDialogStateEvent,
  ArtifactTab,
  ArtifactValidationResult,
  ArtifactProcessingOptions,
} from './types/artifact.types';

// Constants
export { DEFAULT_ARTIFACT_CONFIG, SUPPORTED_MIME_TYPES } from './types/artifact.types';
