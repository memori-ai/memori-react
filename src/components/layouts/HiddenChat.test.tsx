import { render } from '@testing-library/react';
import Memori from '../MemoriWidget/MemoriWidget';
import { integration, memori, tenant } from '../../mocks/data';
import I18nWrapper from '../../I18nWrapper';
import { VisemeProvider } from '../../context/visemeContext';
import { ArtifactSystemProvider } from '../MemoriArtifactSystem';
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // Deprecated
    removeListener: jest.fn(), // Deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

it('renders HIDDEN_CHAT layout unchanged', () => {
  const { container } = render(
    <I18nWrapper>
      <ArtifactSystemProvider config={{
        supportedMimeTypes: {
          html: {
            name: 'HTML',
            icon: '🌐',
            hasPreview: true,
            language: 'html',
            mimeType: 'text/html',
          },
        },
      }}>
      <VisemeProvider>
        <Memori
          showShare={true}
          showSettings={true}
          memori={memori}
          tenant={tenant}
          tenantID="aisuru.com"
          integration={integration}
          layout="HIDDEN_CHAT"
        />
      </VisemeProvider>
      </ArtifactSystemProvider>
    </I18nWrapper>
  );
  expect(container).toMatchSnapshot();
});
