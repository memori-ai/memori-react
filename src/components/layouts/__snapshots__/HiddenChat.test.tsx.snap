// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders HIDDEN_CHAT layout unchanged 1`] = `
<div>
  <div
    class="memori memori-widget memori-layout-hidden_chat memori-controls-bottom memori--avatar-readyplayerme memori--with-integration"
    data-memori-engine-state="{}"
    data-memori-id="66b4e161-2431-4b21-9b70-d8c27de730ca"
    data-memori-integration="cb3c4776-7f0b-4f97-a773-c32a5d7a3bf1"
    data-memori-name="Memori"
    data-memori-secondary-id="25ced51c-3520-41af-8bbe-222d861b8e32"
    style="height: 100vh;"
  >
    <input
      class="memori-sidebar-toggle"
      id="memori-sidebar-toggle"
      type="checkbox"
    />
    <div
      class="memori-sidebar-container"
    >
      <label
        class="memori-sidebar-toggle-label memori-open-label"
        for="memori-sidebar-toggle"
      >
        <svg
          aria-hidden="true"
          class="memori-icon"
          fill="none"
          focusable="false"
          role="img"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <path
            d="M9.09 9a3 3 0 015.83 1c0 2-3 3-3 3"
          />
          <path
            d="M12 17L12.01 17"
          />
        </svg>
      </label>
      <aside
        class="memori-sidebar"
      >
        <label
          class="memori-sidebar-toggle-label memori-close-label"
          for="memori-sidebar-toggle"
        >
          <span>
            <svg
              aria-hidden="true"
              class="memori-icon-close"
              focusable="false"
              role="img"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
              />
            </svg>
          </span>
        </label>
        <div
          class="memori-sidebar-content"
        >
          <div
            class="memori-hidden-chat-layout--header"
          >
            <div
              class="memori-header memori-hidden-chat-layout-header--layout"
            >
              <div
                class="memori-export-history-button memori-header--button memori-header--button--export"
                data-headlessui-state=""
                title="Vuoi esportare la cronologia chat?"
              >
                <button
                  class="memori-button memori-button--circle memori-button--icon-only memori-button--disabled"
                  data-headlessui-state=""
                  disabled=""
                  id="headlessui-popover-button-:r0:"
                  type="button"
                >
                  <div
                    class="memori-button--icon"
                  >
                    <svg
                      aria-hidden="true"
                      focusable="false"
                      role="img"
                      viewBox="0 0 1024 1024"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M464 512a48 48 0 1096 0 48 48 0 10-96 0zm200 0a48 48 0 1096 0 48 48 0 10-96 0zm-400 0a48 48 0 1096 0 48 48 0 10-96 0zm661.2-173.6c-22.6-53.7-55-101.9-96.3-143.3a444.35 444.35 0 00-143.3-96.3C630.6 75.7 572.2 64 512 64h-2c-60.6.3-119.3 12.3-174.5 35.9a445.35 445.35 0 00-142 96.5c-40.9 41.3-73 89.3-95.2 142.8-23 55.4-34.6 114.3-34.3 174.9A449.4 449.4 0 00112 714v152a46 46 0 0046 46h152.1A449.4 449.4 0 00510 960h2.1c59.9 0 118-11.6 172.7-34.3a444.48 444.48 0 00142.8-95.2c41.3-40.9 73.8-88.7 96.5-142 23.6-55.2 35.6-113.9 35.9-174.5.3-60.9-11.5-120-34.8-175.6zm-151.1 438C704 845.8 611 884 512 884h-1.7c-60.3-.3-120.2-15.3-173.1-43.5l-8.4-4.5H188V695.2l-4.5-8.4C155.3 633.9 140.3 574 140 513.7c-.4-99.7 37.7-193.3 107.6-263.8 69.8-70.5 163.1-109.5 262.8-109.9h1.7c50 0 98.5 9.7 144.2 28.9 44.6 18.7 84.6 45.6 119 80 34.3 34.3 61.3 74.4 80 119 19.4 46.2 29.1 95.2 28.9 145.8-.6 99.6-39.7 192.9-110.1 262.7z"
                      />
                    </svg>
                  </div>
                </button>
              </div>
              <button
                class="memori-button memori-button--primary memori-button--circle memori-button--padded memori-button--icon-only memori-header--button memori-header--button-settings"
                title="Impostazioni"
              >
                <span
                  class="memori-button--icon"
                >
                  <svg
                    aria-hidden="true"
                    focusable="false"
                    role="img"
                    viewBox="0 0 1024 1024"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M924.8 625.7l-65.5-56c3.1-19 4.7-38.4 4.7-57.8s-1.6-38.8-4.7-57.8l65.5-56a32.03 32.03 0 009.3-35.2l-.9-2.6a443.74 443.74 0 00-79.7-137.9l-1.8-2.1a32.12 32.12 0 00-35.1-9.5l-81.3 28.9c-30-24.6-63.5-44-99.7-57.6l-15.7-85a32.05 32.05 0 00-25.8-25.7l-2.7-.5c-52.1-9.4-106.9-9.4-159 0l-2.7.5a32.05 32.05 0 00-25.8 25.7l-15.8 85.4a351.86 351.86 0 00-99 57.4l-81.9-29.1a32 32 0 00-35.1 9.5l-1.8 2.1a446.02 446.02 0 00-79.7 137.9l-.9 2.6c-4.5 12.5-.8 26.5 9.3 35.2l66.3 56.6c-3.1 18.8-4.6 38-4.6 57.1 0 19.2 1.5 38.4 4.6 57.1L99 625.5a32.03 32.03 0 00-9.3 35.2l.9 2.6c18.1 50.4 44.9 96.9 79.7 137.9l1.8 2.1a32.12 32.12 0 0035.1 9.5l81.9-29.1c29.8 24.5 63.1 43.9 99 57.4l15.8 85.4a32.05 32.05 0 0025.8 25.7l2.7.5a449.4 449.4 0 00159 0l2.7-.5a32.05 32.05 0 0025.8-25.7l15.7-85a350 350 0 0099.7-57.6l81.3 28.9a32 32 0 0035.1-9.5l1.8-2.1c34.8-41.1 61.6-87.5 79.7-137.9l.9-2.6c4.5-12.3.8-26.3-9.3-35zM788.3 465.9c2.5 15.1 3.8 30.6 3.8 46.1s-1.3 31-3.8 46.1l-6.6 40.1 74.7 63.9a370.03 370.03 0 01-42.6 73.6L721 702.8l-31.4 25.8c-23.9 19.6-50.5 35-79.3 45.8l-38.1 14.3-17.9 97a377.5 377.5 0 01-85 0l-17.9-97.2-37.8-14.5c-28.5-10.8-55-26.2-78.7-45.7l-31.4-25.9-93.4 33.2c-17-22.9-31.2-47.6-42.6-73.6l75.5-64.5-6.5-40c-2.4-14.9-3.7-30.3-3.7-45.5 0-15.3 1.2-30.6 3.7-45.5l6.5-40-75.5-64.5c11.3-26.1 25.6-50.7 42.6-73.6l93.4 33.2 31.4-25.9c23.7-19.5 50.2-34.9 78.7-45.7l37.9-14.3 17.9-97.2c28.1-3.2 56.8-3.2 85 0l17.9 97 38.1 14.3c28.7 10.8 55.4 26.2 79.3 45.8l31.4 25.8 92.8-32.9c17 22.9 31.2 47.6 42.6 73.6L781.8 426l6.5 39.9zM512 326c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm79.2 255.2A111.6 111.6 0 01512 614c-29.9 0-58-11.7-79.2-32.8A111.6 111.6 0 01400 502c0-29.9 11.7-58 32.8-79.2C454 401.6 482.1 390 512 390c29.9 0 58 11.6 79.2 32.8A111.6 111.6 0 01624 502c0 29.9-11.7 58-32.8 79.2z"
                    />
                  </svg>
                </span>
              </button>
              <div
                class="memori-share-button memori-share-button--align-left"
                data-headlessui-state=""
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="memori-button memori-button--circle memori-button--icon-only memori-share-button--button memori-header--button memori-header--button-share memori-button--primary"
                  data-headlessui-state=""
                  id="headlessui-menu-button-:r5:"
                  title="Condividi"
                  type="button"
                >
                  <div
                    class="memori-button--icon"
                  >
                    <svg
                      aria-hidden="true"
                      focusable="false"
                      role="img"
                      viewBox="0 0 1024 1024"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M752 664c-28.5 0-54.8 10-75.4 26.7L469.4 540.8a160.68 160.68 0 0 0 0-57.6l207.2-149.9C697.2 350 723.5 360 752 360c66.2 0 120-53.8 120-120s-53.8-120-120-120-120 53.8-120 120c0 11.6 1.6 22.7 4.7 33.3L439.9 415.8C410.7 377.1 364.3 352 312 352c-88.4 0-160 71.6-160 160s71.6 160 160 160c52.3 0 98.7-25.1 127.9-63.8l196.8 142.5c-3.1 10.6-4.7 21.8-4.7 33.3 0 66.2 53.8 120 120 120s120-53.8 120-120-53.8-120-120-120zm0-476c28.7 0 52 23.3 52 52s-23.3 52-52 52-52-23.3-52-52 23.3-52 52-52zM312 600c-48.5 0-88-39.5-88-88s39.5-88 88-88 88 39.5 88 88-39.5 88-88 88zm440 236c-28.7 0-52-23.3-52-52s23.3-52 52-52 52 23.3 52 52-23.3 52-52 52z"
                      />
                    </svg>
                  </div>
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          id="extension"
        />
        <div
          class="memori-hidden-chat-layout--controls"
        >
          <div
            class="memori-loading"
          >
            <div
              class="memori-spin"
            >
              <div
                class="memori-spin--spinner"
              >
                <svg
                  aria-hidden="true"
                  class="memori-loading-icon"
                  focusable="false"
                  role="img"
                  viewBox="0 0 1024 1024"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </aside>
    </div>
    <audio
      id="memori-audio"
      src="https://aisuru.com/intro.mp3"
      style="display: none;"
    />
  </div>
  <script
    async=""
    id="MathJax-script"
    src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"
  />
</div>
`;
